rslib/src/import_export/mod.rs

```diff
@@ -1,10 +1,98 @@
 // Copyright: Ankitects Pty Ltd and contributors
 // License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 
+mod gather;
+mod insert;
 pub mod package;
 
+use crate::prelude::*;
+
 #[derive(Debug, Clone, Copy, PartialEq)]
 pub enum ImportProgress {
-    Collection,
+    File,
     Media(usize),
+    MediaCheck(usize),
+    Notes(usize),
+}
+
+/// Wrapper around a progress function, usually passed by the [crate::backend::Backend],
+/// to make repeated calls more ergonomic.
+pub(crate) struct IncrementableProgress<P> {
+    progress_fn: Box<dyn FnMut(P, bool) -> bool>,
+    count_map: Option<Box<dyn FnMut(usize) -> P>>,
+    count: usize,
```

Comment:
I was a bit concerned about the extra overhead of Box<dyn>, as we may be invoking increment() in a hot loop. Whether it actually matters or not in our current code I don't know - the other stuff we're doing in the loop like SQL lookups or sha1 calculations probably dwarfs it.

One possible alternative would be to change set_count_map() to zero(), get rid of count_map, and pass an impl into increment instead. But either is ok with me.

Reply:
Yeah, with all the I/O and allocations going on, it's hard to imagine that some derefs will carry weight.
But I added a new struct that is generic over `count_map`. That's an improvement with regards to the API, because users can't forget to call `set_count_map()` first.
However, I don't see how we can ergonomically use a generic for `progress_fn`, because we'd like to pass that struct around, and each time annotating it with `&mut IncrementableProgress<ImportProgress, impl FnMut(P, bool) -> bool>` seems pretty cumbersome. Here's hoping that `impl`s in type aliases will be stabilised at some point.

Reply:
That seems like a good compromise - thanks Rumo. 

>  Here's hoping that impls in type aliases will be stabilised at some point.

Agreed!